File: MEPmodel_bio_core.m
function sim = MEPmodel_bio_core(model)

    DIwave         = model.DIwaveConv;
    t              = model.t;
    muaps          = model.muaps;
    tmuap          = model.tmuap;
    R              = model.R;      % membrane resistance of MNs
    Wexc           = model.Wexc;
    RWinh          = model.RWinh;   % R.*Winh
    dt             = model.dt;
    tauLIF         = model.tauLIF;
    Eexc           = model.Eexc;
    Einh           = model.Einh;
    V_rest         = model.V_rest;
    V_thr          = model.V_thr;
    T_ref          = model.T_ref;
    Tmu            = model.Tmu;
    withRC         = model.withRC;
    maxES          = model.maxES;
    fastAChRweight = model.fastAChRweight;
    rc             = model.rc;
    tau            = model.kernel.tau;
    h              = model.kernel.h;

    %------------record for all TMS intensities--------------
    nIntensities=size(DIwave,1);
    simMEP=zeros(nIntensities,length(t));
    spike_times=nan(100,maxES,nIntensities); 
    spike_times2=cell(nIntensities,1);         % all spike times

    gexc_all   = zeros(nIntensities,length(t)); % (for record) 
    ginh_all  = zeros(nIntensities,length(t)); % (for record) 
    Rm_all  = zeros(nIntensities,length(t)); % (for record) 
    Vm_all  = zeros(nIntensities,length(t)); % (for record) 
    mRC_all = zeros(nIntensities,length(t)); % RC rate (for record) 
    mMN_all = zeros(nIntensities,length(t)); % MN rate (for record) 
    vRC_all = zeros(nIntensities,length(t)); % EPSP (MN->RC) (for record)
    Iexc_all = zeros(nIntensities,length(t)); % EPSP (MN->RC) (for record)
    Iinh_all = zeros(nIntensities,length(t)); % EPSP (MN->RC) (for record)

    for i=1:nIntensities
       
        %----- Initialization -----
        gexc   = DIwave(i,:);                  % synaptic conductances on MNs
        ginh   = zeros(1,length(t));           % synaptic conductances on MNs
        Iexc   = zeros(100,length(t));         % synaptic current to MNs
        Iinh   = zeros(100,length(t));         % synaptic current to MNs
        Vm     = ones(100,length(t)) * V_rest; % MN membrane potentials
        Vm_lag = ones(100,length(t)) * V_rest; % MN membrane potentials (a copy for interpolation)
        sMN    = zeros(100,length(t));         % MN spikes (0/1) 
        TR     = zeros(100,1);                 % Timer of refractory period
        mMN    = zeros(1,length(t));           % MN firing rate
        mRC    = zeros(1,length(t));           % RC firing rate   
        v      = zeros(3,length(t));           % average postsynaptic potential
        v2     = zeros(3,length(t));           %   EPSPs (MN->RC): fast AchR (rise 0.5 ms, decay 3.6 ms) 
                                               %                   slow AchR (rise 1.8 ms, decay 20.2 ms)  
                                               %   IPSP (RC->MN):  GlyR (rise 1 ms, decay 6 ms) 
           
        %----- simulation -----
        for tt=1:length(t)-1

            %-----MN rate-----
            mMN(tt) = mean(Wexc'.*sMN(:,tt)); % mean MN firing rate 

            %-----RC rate-----
            if withRC
                rescale = 100; %N=100
                mRC(tt) = sigmoid(v(1,tt)*fastAChRweight + v(2,tt)*(1-fastAChRweight), ...
                                  rc.v_thr/rescale, rc.r*rescale, rc.fmax);
            end
            %-----PSPs (biexponential kernel)----- 
            in = [mMN(tt); % (MN->RC, fast AchR)
                  mMN(tt); % (MN->RC, slow AchR)
                  mRC(tt)];% (RC->MN, GlyR)

            dv  = v2(:,tt);
            dv2 = h.*in - ...
                  v2(:,tt).*(tau(:,1)+tau(:,2))./tau(:,1)./tau(:,2) - ...
                  v(:,tt)./tau(:,1)./tau(:,2); 

            v(:,tt+1)  = v(:,tt)  + dv*dt;
            v2(:,tt+1) = v2(:,tt) + dv2*dt;

            
            if withRC 
                ginh(tt)  =v(3,tt);
                Iexc(:,tt)=(Eexc-Vm(:,tt))*gexc(tt);         
                Iinh(:,tt)=(Einh-Vm(:,tt))*ginh(tt);
                Vm(:,tt+1) = Vm(:,tt) + ...
                             dt / tauLIF * (V_rest - Vm(:,tt) + ...
                                            R'.*Iexc(:,tt) + ... % excitatory input (from DI-waves)
                                            RWinh'.*Iinh(:,tt));  % inhibitory input (from RC)
            else % no RC 
                Iexc(:,tt)=(Eexc-Vm(:,tt))*gexc(tt);
                Vm(:,tt+1) = Vm(:,tt) + ...
                             dt / tauLIF * (V_rest - Vm(:,tt) + ...
                                            R'.*Iexc(:,tt)); % excitatory input (from DI-waves)
            end
            %-----refractory period-----
            idx = TR>0;
            Vm(idx,tt+1) = V_rest; % rest
            TR(idx,1)=TR(idx,1)-1;
    
            %-----spike of 100 motor neurons----- 
            idx = Vm(:,tt+1) > V_thr;
            Vm_lag(idx,tt+1) = Vm(idx,tt+1); % just to record the Vm right before spiking (for time interpolation)
            sMN(idx,tt+1) = 1; % spike
            Vm(idx,tt+1) = V_rest; % rest
            TR(idx,1)=round(T_ref/dt);
    
        end
        
        %----- spike times -----  
        MEPcomps = zeros(100,length(t));
        T=Tmu/dt;

        for n=1:100
            [~,idx]=find(Vm_lag(n,:)> V_thr);  % all spikes
            %-----collect effective spikes-----
            idxES=nan(1,maxES); 
            refractory_idx=0;
            counter=1;
            while ~isempty(idx) && counter<=maxES 
                if idx(1)>refractory_idx
                  idxES(counter)=idx(1);
                  refractory_idx=idx(1)+T;
                  counter=counter+1;
                end
                idx(1)=[];
            end
            idxES(isnan(idxES))=[];
            %---------------------------------
            for k=1:length(idxES) 
                spike_times(n,k,i) = dt/(Vm_lag(n,idxES(k))-Vm(n,idxES(k)-1)) *(V_thr-Vm(n, idxES(k)-1))+ t(idxES(k)-1);
                MEPcomps(n,:) = MEPcomps(n,:) + interp1(tmuap+spike_times(n,k,i),muaps(:,n),t,'linear',0); %cumulative
            end          
        end
        simMEP(i,:) = sum(MEPcomps,1); % [10 x  500]

        %---------for record--------
        gexc_all(i,:)=R(1)*gexc;  % conductance (Cortex->MN1)    
        Rm_all(i,:)=sMN(1,:); % spike of the first MN
        Vm_all(i,:)=Vm(1,:); % spike of the first MN
        [tmp1,tmp2]=find(sMN);
        spike_times2{i}=[tmp1,t(tmp2)']; 


        mMN_all(i,:)=mMN(:); % MN rate
        vRC_all(i,:)=v(1,:)*fastAChRweight+v(2,:)*(1-fastAChRweight); % EPSP (MN->RC)
        Iexc_all(i,:)=Iexc(1,:); % excitatoty input to MN1 (from DIwaves)
        Iinh_all(i,:)=Iinh(1,:); % inhibitory input to MN1 (from RC)

        if withRC
            mRC_all(i,:)=mRC(:); % RC rate
            ginh_all(i,:)=RWinh(1)*ginh;%v(3,:); % conductance (RC->MN1)
            
        end
    end

    sim.t=t;
    sim.simMEP=simMEP;
    sim.spike_times=spike_times;
    sim.gexc_all=gexc_all;
    sim.ginh_all=ginh_all;
    sim.Rm_all=Rm_all;
    sim.Vm_all=Vm_all;
    sim.mRC_all=mRC_all;
    sim.mMN_all=mMN_all;
    sim.vRC_all=vRC_all;
    sim.spike_times2=spike_times2;
    sim.Iexc_all=Iexc_all;
    sim.Iinh_all=Iinh_all;
end

File: MEPmodel_bio_core.py
import numpy as np
from sigmoid import sigmoid
from scipy.interpolate import interp1d

def MEPmodel_bio_core(model):
    # Mapping dictionary keys from the model structure
    DIwave         = model['DIwaveConv']
    t              = model['t']
    muaps          = model['muaps']
    tmuap          = model['tmuap']
    R              = model['R']             # membrane resistance of MNs
    Wexc           = model['Wexc']
    RWinh          = model['RWinh']         # R * Winh
    dt             = model['dt']
    tauLIF         = model['tauLIF']
    Eexc           = model['Eexc']
    Einh           = model['Einh']
    V_rest         = model['V_rest']
    V_thr          = model['V_thr']
    T_ref          = model['T_ref']
    Tmu            = model['Tmu']
    withRC         = model['withRC']
    maxES          = model['maxES']
    fastAChRweight = model['fastAChRweight']
    rc             = model['rc']
    tau            = np.array(model[('kernel', 'tau')]) # Nested structure handled as dict
    h              = np.array(model[('kernel', 'h')])

    # ------------record for all TMS intensities--------------
    nIntensities = DIwave.shape[0]
    simMEP = np.zeros((nIntensities, len(t)))
    spike_times = np.full((100, maxES, nIntensities), np.nan)
    spike_times2 = [None] * nIntensities      # all spike times (list of objects)

    gexc_all = np.zeros((nIntensities, len(t)))
    ginh_all = np.zeros((nIntensities, len(t)))
    Rm_all   = np.zeros((nIntensities, len(t)))
    Vm_all   = np.zeros((nIntensities, len(t)))
    mRC_all  = np.zeros((nIntensities, len(t)))
    mMN_all  = np.zeros((nIntensities, len(t)))
    vRC_all  = np.zeros((nIntensities, len(t)))
    Iexc_all = np.zeros((nIntensities, len(t)))
    Iinh_all = np.zeros((nIntensities, len(t)))

    for i in range(nIntensities):
       
        # ----- Initialization -----
        gexc   = DIwave[i, :]                  # synaptic conductances on MNs
        ginh   = np.zeros(len(t))              # synaptic conductances on MNs
        Iexc   = np.zeros((100, len(t)))       # synaptic current to MNs
        Iinh   = np.zeros((100, len(t)))       # synaptic current to MNs
        Vm     = np.ones((100, len(t))) * V_rest 
        Vm_lag = np.ones((100, len(t))) * V_rest 
        sMN    = np.zeros((100, len(t)))       # MN spikes (0/1) 
        TR     = np.zeros(100)                 # Timer of refractory period
        mMN    = np.zeros(len(t))              # MN firing rate
        mRC    = np.zeros(len(t))              # RC firing rate   
        v      = np.zeros((3, len(t)))         # average postsynaptic potential
        v2     = np.zeros((3, len(t)))         # derivative helper
           
        # ----- simulation -----
        for tt in range(len(t) - 1):

            # -----MN rate-----
            mMN[tt] = np.mean(Wexc * sMN[:, tt]) # Wexc' in MATLAB is handled by element-wise if Wexc is 1D

            # -----RC rate-----
            if withRC:
                rescale = 100 
                # Assuming sigmoid is defined elsewhere as a custom function
                mRC[tt] = sigmoid(v[0, tt] * fastAChRweight + v[1, tt] * (1 - fastAChRweight), 
                                  rc['v_thr'] / rescale, rc['r'] * rescale, rc['fmax'])
            
            # -----PSPs (biexponential kernel)----- 
            input_vec = np.array([mMN[tt], mMN[tt], mRC[tt]])

            dv  = v2[:, tt]
            # print("shape tau ", np.shape(tau))
            # print("shape input vec ", np.shape(input_vec))
            # print("shape h ", np.shape(h))
            # print("shape v ", np.shape(v))
            # print("shape v2 ", np.shape(v2))
            # print("tt", tt)

            h = h.flatten()


            dv2 = (h * input_vec - v2[:, tt] * (tau[:, 0] + tau[:, 1]) / (tau[:, 0] * tau[:, 1]) - v[:, tt] / (tau[:, 0] * tau[:, 1]))

            v[:, tt + 1]  = v[:, tt]  + dv * dt
            v2[:, tt + 1] = v2[:, tt] + dv2 * dt

            # print("shape dv ", np.shape(dv))
            # print("shape dv2 ", np.shape(dv2))




            if withRC: 
                ginh[tt] = v[2, tt]
                Iexc[:, tt] = (Eexc - Vm[:, tt]) * gexc[tt]         
                Iinh[:, tt] = (Einh - Vm[:, tt]) * ginh[tt]
                Vm[:, tt + 1] = Vm[:, tt] + (dt / tauLIF) * (V_rest - Vm[:, tt] + 
                                R * Iexc[:, tt] + RWinh * Iinh[:, tt])
            else: 
                Iexc[:, tt] = (Eexc - Vm[:, tt]) * gexc[tt]
                Vm[:, tt + 1] = Vm[:, tt] + (dt / tauLIF) * (V_rest - Vm[:, tt] + 
                                R * Iexc[:, tt])

            # -----refractory period-----
            idx_tr = TR > 0
            Vm[idx_tr, tt + 1] = V_rest
            TR[idx_tr] -= 1
    
            # -----spike of 100 motor neurons----- 
            idx_spike = Vm[:, tt + 1] > V_thr
            Vm_lag[idx_spike, tt + 1] = Vm[idx_spike, tt + 1] 
            sMN[idx_spike, tt + 1] = 1 
            Vm[idx_spike, tt + 1] = V_rest 
            TR[idx_spike] = round(T_ref / dt)
    
        # ----- spike times -----  
        MEPcomps = np.zeros((100, len(t)))
        T_idx = Tmu / dt

        for n in range(100):
            # Find all spikes for neuron n
            idx = np.where(Vm_lag[n, :] > V_thr)[0]
            
            # -----collect effective spikes-----
            idxES = []
            refractory_idx = 0
            counter = 0
            
            while len(idx) > 0 and counter < maxES:
                if idx[0] > refractory_idx:
                    idxES.append(idx[0])
                    refractory_idx = idx[0] + T_idx
                    counter += 1
                idx = idx[1:]
            
            # ---------------------------------
            for k in range(len(idxES)):
                curr_idx = idxES[k]
                # Linear interpolation for precise spike time
                spike_t = (dt / (Vm_lag[n, curr_idx] - Vm[n, curr_idx - 1]) * (V_thr - Vm[n, curr_idx - 1]) + t[curr_idx - 1])
                
                spike_times[n, k, i] = spike_t
                
                # Cumulative interpolation of MUAP
                f_interp = interp1d((tmuap + spike_t).flatten(), (muaps[:, n]).flatten(), kind='linear', 
                                    bounds_error=False, fill_value=0)
                MEPcomps[n, :] += f_interp(t)
        
        simMEP[i, :] = np.sum(MEPcomps, axis=0)

        # ---------for record--------
        gexc_all[i, :] = R[0] * gexc  
        Rm_all[i, :]   = sMN[0, :] 
        Vm_all[i, :]   = Vm[0, :] 
        
        spike_rows, spike_cols = np.where(sMN == 1)
        spike_times2[i] = np.column_stack((spike_rows + 1, t[spike_cols])) 

        mMN_all[i, :]  = mMN
        vRC_all[i, :]  = v[0, :] * fastAChRweight + v[1, :] * (1 - fastAChRweight)
        Iexc_all[i, :] = Iexc[0, :]
        Iinh_all[i, :] = Iinh[0, :]

        if withRC:
            mRC_all[i, :]  = mRC
            ginh_all[i, :] = RWinh[0] * ginh

    # Construct output dictionary
    sim = {
        't': t,
        'simMEP': simMEP,
        'spike_times': spike_times,
        'gexc_all': gexc_all,
        'ginh_all': ginh_all,
        'Rm_all': Rm_all,
        'Vm_all': Vm_all,
        'mRC_all': mRC_all,
        'mMN_all': mMN_all,
        'vRC_all': vRC_all,
        'spike_times2': spike_times2,
        'Iexc_all': Iexc_all,
        'Iinh_all': Iinh_all
    }
    return sim